### PR TITLE
feat: Collect agent SSH metrics

### PR DESCRIPTION
**Pull request by mtojek**
Related: https://github.com/coder/coder/issues/6724

This PR adds more metrics to the SSH agent. These should help with identifying connectivity issues. 

~~Actually, it is similar to metrics support in the [tailssh](https://github.com/tailscale/tailscale/blob/f0ee03dfafda9c5a8804b25f20ea881090bd4ab5/ssh/tailssh/tailssh.go#L1830-L1842).~~ challenged by Spike to use metric labels.

Sample metrics:
```
agent_sessions_total{agent_name="main",magic_type="ssh",pty="yes",username="admin",workspace_name="docker-1"} 5
agent_sessions_total{agent_name="main",magic_type="vscode",pty="no",username="admin",workspace_name="docker-1"} 1

```

### CODE REVIEW

**Review by spikecurtis — COMMENTED** [review]
I'm wary of using global state for this.

We originally did it this way with `promauto` in V1 and ripped it out later.

https://github.com/coder/v1/issues/13357

Global state is basically impossible to test and can create unintended side effects just by importing packages.

I realize we're already a bit wed to `tailscale.com/util/clientmetric` in order to get metrics out of tailscale, but I don't think we should add to the problem.

As an alternative, we could attach a prometheus registry to the agent and register the metrics there, and then collect them to ship to coderd.

**Comment by mtojek** [comment]
I refactored the source to use PrometheusRegistry. It resulted in implementing a custom handler to merge Tailscale and agent metrics. Let me know your thoughts, and in the meantime, I will proceed with testing.

Unfortunately, the size of PR jumped from ~160 to ~450 LOC. I guess that I can limit it but providing some wrapper for `prometheus.NewCounter` + `registerer.MustRegister`.

File: agent/agentssh/agentssh_internal_test.go
Comment (spikecurtis) [inline agent/agentssh/agentssh_internal_test.go:64]: this should be above the comment so that it's more clear the comment is referring to the `s.startPTYSession` line.

File: agent/agentssh/agentssh.go
Comment (spikecurtis) [inline agent/agentssh/agentssh.go:378]: note that my earlier comment about wrapping errors and incrementing the metrics higher in the stack do *not* apply to these errors that are in separate goroutines---there is nothing higher in the stack to handle incrementing metrics, and in a real sense it is a distinct error from any errors hit on the main goroutine, so we're not double counting by incrementing here.

File: cli/agent.go
Comment (spikecurtis) [inline cli/agent.go:404]: 👍

File: agent/metrics.go
Comment (spikecurtis) [inline agent/metrics.go:105]: :+1:

File: agent/metrics.go
Comment (spikecurtis) [inline agent/metrics.go:108]: Oh man, it's a real shame we never thought to include labels on metrics in the first place.

I think we should.  They're just too useful to ignore and we publish on the coderd side with prometheus.

File: agent/agentssh/metrics.go
Comment (mtojek) [reply agent/agentssh/metrics.go:55]: I can rephrase them, but I left the reference to "callbacks" on purpose. I didn't assume that these metrics should be considered pure error statistics but rather control checks to debug the execution flow. Most likely operators will ignore 90% of them, but the rest can be pretty useful in more advanced investigations.

File: agent/agentssh/agentssh_internal_test.go
Comment (mtojek) [reply agent/agentssh/agentssh_internal_test.go]: Fixed

File: cli/agent.go
Comment (mtojek) [reply cli/agent.go:396]: Fixed on line 390.

File: agent/agentssh/agentssh.go
Comment (spikecurtis) [reply agent/agentssh/agentssh.go:212]: In prometheus it's easy to query for the sum across the label of error metrics.  No chance of missing one if more label values are added later.

File: agent/metrics.go
Comment (mtojek) [reply agent/metrics.go:108]: Let me share some background - in the [original implementation](https://github.com/coder/coder/pull/7183) I focused on passing the magicsock metrics that are pretty flat (no vectors, no labels). I didn't have a reason to implement an overkill with unused labels (KISS).

This is the first time on the agent side that requires support for metric vectors (and labels). As you noticed, it means that we need to improve the receiver side too (coderd).

I'm not quite sure if I should do this as part of this PR, or as a separate story, and refactor metrics to use labels.

File: agent/agentssh/metrics.go
Comment (mtojek) [reply agent/agentssh/metrics.go]: I removed all callbacks, and renamed `connectionFailedCallback` to `failedConnectionsTotal`.

File: agent/metrics.go
Comment (mtojek) [reply agent/metrics.go]: Fixed

File: agent/agentssh/metrics.go
Comment (mtojek) [reply agent/agentssh/metrics.go]: Refactored

File: agent/agentssh/metrics.go
Comment (mtojek) [reply agent/agentssh/metrics.go]: Done

File: agent/agentssh/agentssh.go
Comment (johnstcn) [inline agent/agentssh/agentssh.go:230]:
```suggestion
```

**Review by johnstcn — APPROVED** [review]

File: agent/agentssh/agentssh.go
Comment (mtojek) [reply agent/agentssh/agentssh.go]: Removed unused function. Thanks

File: agent/metrics.go
Comment (mtojek) [reply agent/metrics.go]: Added support for aggregating metrics by labels.

**Comment by mtojek** [comment]
@spikecurtis I refactored the source code to use metric labels. Looking at the code now (although it is still a draft), I'm not convinced about the error-wrapping technique. 

Let me challenge it. There are cases where we'd like to collect more than one error, and trying to aggregate a few of them requires extra logic (at least `append` operation).

Consider the following code:

```go
	if !isQuietLogin(session.RawCommand()) {
		manifest := s.Manifest.Load()
		if manifest != nil {
			err := showMOTD(session, manifest.MOTDFile)
			if err != nil {
				s.logger.Error(ctx, "show MOTD", slog.Error(err))
				s.metrics.sessionErrors.WithLabelValues(magicTypeLabel, "yes", "motd").Add(1)
			}
		} else {
			s.logger.Warn(ctx, "metadata lookup failed, unable to show MOTD")
		}
	}
```

We'd like to know that something went wrong with MOTD, but it doesn't mean that the session gets terminated instantly. If something bad happens to the session later, the code will have to bubble up a sum of errors. Personally, I don't see a big gain here compared to a simple intercepting one-liner:

```go
s.metrics.sessionErrors.WithLabelValues(magicTypeLabel, "yes", "motd").Add(1)
```

File: agent/agent_test.go
Comment (johnstcn) [inline agent/agent_test.go:1729]: suggestion: remove `ptyTimeout` and set a default value in `options`

File: agent/agent_test.go
Comment (johnstcn) [inline agent/agent_test.go:2093]: Use `assert` instead of `require` in `Eventually()` as this amounts to calling `t.FailNow()` in a goroutine, which is unsafe.

File: coderd/prometheusmetrics/aggregator.go
Comment (johnstcn) [inline coderd/prometheusmetrics/aggregator.go:151]: nit: rename this to be more descriptive; it appears to serve an important function.

File: coderd/prometheusmetrics/aggregator_test.go
Comment (johnstcn) [inline coderd/prometheusmetrics/aggregator_test.go:120]: Adding `d.String()` to the `require()` call here is probably not a bad idea here too.

File: agent/agent_test.go
Comment (johnstcn) [inline agent/agent_test.go:2076]: `t.Helper()`

**Review by johnstcn — COMMENTED** [review]
I only have nits, but will defer to others for approval.

File: coderd/prometheusmetrics/aggregator.go
Comment (mtojek) [reply coderd/prometheusmetrics/aggregator.go:151]: I'm not sure if I know what you mean. Are you referring to the variable `i`? I can rename it to `labelIndex`, I guess?

File: agent/agent_test.go
Comment (mtojek) [reply agent/agent_test.go:2077]: Fixed

File: coderd/prometheusmetrics/aggregator_test.go
Comment (mtojek) [reply coderd/prometheusmetrics/aggregator_test.go]: Fixed

File: coderd/prometheusmetrics/aggregator.go
Comment (johnstcn) [reply coderd/prometheusmetrics/aggregator.go:151]: Yes, correct. `labelIndex` would be fine!

**Comment by spikecurtis** [comment]
> I'm not convinced about the error-wrapping technique.
> 
> Let me challenge it. There are cases where we'd like to collect more than one error, and trying to aggregate a few of them requires extra logic (at least `append` operation).

That's fair---the high level goal is that _each legitimately distinct error is counted exactly once_.  For errors that don't terminate the session, I think it's fine to just directly increment the error counter and move on.

What I want to avoid though, is double-counting, where an error type is incremented, and then a "generic error" count is also incremented after the function returns.  We could limit the error wrapping to just those errors that trigger an immediate return in the session handler.  The feature I really like about error wrapping is that it affords us the ability to ensure that we really are counting all the session-ending errors and not just forgetting in some `return` paths.

File: coderd/prometheusmetrics/aggregator.go
Comment (spikecurtis) [inline coderd/prometheusmetrics/aggregator.go:179]: I think it would be clearer to set both the label name and value in a single pass through `m.Labels`

File: coderd/prometheusmetrics/aggregator.go
Comment (spikecurtis) [inline coderd/prometheusmetrics/aggregator.go:183]: I'd suggest pulling this entire procedure out into a method on `annotatedMetric`, eg

```
constMetric, err := m.asPrometheus()
```

**Comment by mtojek** [comment]
I see your point, but considering the fact that we have to increase the error counter in two different ways, the argumentation does not convince me.

> What I want to avoid though, is double-counting, where an error type is incremented, and then a "generic error" count is also incremented after the function returns.

It is a theoretical use case, as most crucial code paths fail fast and return (connection is terminated). There is a low chance that the same counter is incremented twice.

> The feature I really like about error wrapping is that it affords us the ability to ensure that we really are counting all the session-ending errors and not just forgetting in some return paths.

This is a valid point, but in my opinion, it complicates the code logic too much - I suppose that this is a matter of personal preference.

I'm wondering if we can push this change in the follow-up iteration to deliver the feature first. What you're suggesting here is pure refactoring. Nevertheless, I will start doing it now 👍 .

File: coderd/prometheusmetrics/aggregator_test.go
Comment (spikecurtis) [inline coderd/prometheusmetrics/aggregator_test.go:49]: this feels fragile.

Your tests could build string maps of the expected and actual labels and then `require.Equal` on the maps.

**Comment by spikecurtis** [comment]
> It is a theoretical use case, as most crucial code paths fail fast and return (connection is terminated). There is a low chance that the same counter is incremented twice.

I don't mean as some race condition --- I just mean counting an error at two different places in the stack.  Your earlier PR incremented metrics both in `sessionStart` and then also in `sessionHandler` if it returned an error.  Doesn't look like this PR does this any more, so it's fine not to implement error wrapping if you feel strongly about it.

File: agent/agentssh/agentssh.go
Comment (spikecurtis) [inline agent/agentssh/agentssh.go:250]: this metric takes 3 label values

File: agent/agentssh/agentssh.go
Comment (spikecurtis) [inline agent/agentssh/agentssh.go:257]: metric takes 3 label values

File: codersdk/agentsdk/agentsdk.go
Comment (mtojek) [reply codersdk/agentsdk/agentsdk.go:502]: I prefer using slices over maps to keep the order of elements in transit. I didn't want to sort it specifically on the receiver side.

File: coderd/prometheusmetrics/aggregator.go
Comment (mtojek) [reply coderd/prometheusmetrics/aggregator.go:126]: Refactored

File: agent/agentssh/agentssh.go
Comment (mtojek) [reply agent/agentssh/agentssh.go:250]: Fixed, thanks for catching

File: coderd/prometheusmetrics/aggregator.go
Comment (mtojek) [reply coderd/prometheusmetrics/aggregator.go]: Good idea 👍

File: coderd/prometheusmetrics/aggregator.go
Comment (mtojek) [reply coderd/prometheusmetrics/aggregator.go]: Fixed

File: coderd/prometheusmetrics/aggregator.go
Comment (mtojek) [reply coderd/prometheusmetrics/aggregator.go]: Fixed

File: coderd/prometheusmetrics/aggregator_test.go
Comment (mtojek) [reply coderd/prometheusmetrics/aggregator_test.go:49]: Fixed

**Review by spikecurtis — APPROVED** [review]

**Comment by spikecurtis** [comment]
LGTM!

**Pull request merged**
